lib/src/coap_config.dart

```diff
@@ -83,4 +83,24 @@ abstract class DefaultCoapConfig {
   int get markAndSweepInterval => 10 * 1000; // ms
 
   int get channelReceivePacketSize => 2048;
+
+  /// Whether tinydtls bindings via the [tinydtls] package should be used for
+  /// CoAPS.
+  bool get dtlsUseTinyDtls => false;
```

Comment:
Could we perhaps handle this similar to how version is handled? For example:
```dart
String get dtlsBackend => 'TinyDtls';
// or
String get dtlsBackend => 'OpenSSL';
```

If I got it right, the user should select one DTLS backend. What should the user expect if `dtlsUseTinyDtls` and `dtlsUseOpenSSL` are both set to `true`? It would be a stupid config (unless I misunderstood), but if the config itself prevents that ambiguity it could make things a little clearer to the end user.

Reply:
Oh, I actually intended both to be usable at the same time as they have different feature sets. Once the OpenSSL binding catches up, tinydtls could then be phased out gradually.

Reply:
But doesn't the user ultimately have to choose one of the two available backends to use with one specific client? Just took a quick glance so I could be wrong, I was just judging by `_determineDtlsNetwork`, which seems to return only a `CoapNetworkTinyDtls` if both TinyDtls and OpenSSL are used at the same time?

I understand that you might want to have one client using TinyDtls and another client using OpenSSL, in which case their client configurations would be different, but using two backends simultaneously to send the same requests is confusing me a bit.

Reply:
Hmm, yeah, the logic is a bit complicated and not well documented, yet. What I intended is the following:

- Assume the user has both tinydtls and OpenSSL enabled. If the user does not provide PSK or ECDSA credentials, then OpenSSL (i.e., certificate mode) is used. If at least one of the two kinds of credentials is present, then tinydtls is used and OpenSSL will be ignored.
- In the case that only tinydtls is loaded, and no credentials are provided, an exception is thrown.
- If only OpenSSL is active, then the client will always use certificates.
- Finally, if no backend is present, another kind of exception is thrown.

So, yes, the user does in fact choose a backend, but only implicitly by either providing credentials or not (after the initial review I thought this might be better to have a more stable API). However, if this should be made more explicit, then I would reintroduce an enum (or a string?) to have users make a conscious decision in this regard.

Reply:
Ah, I understand. I was thinking the developer would consciously decide which library to bundle based on the library's capabilities, since there is some effort involved in setting up and maintaining the bundling. I think OpenSSL also has to be bundled for some platforms (iOS/Android), common targets for Flutter which also happen to be on metered connections where a smaller app size is an advantage. Not sure about Windows, but I think OpenSSL is not shipped with the OS there either?

Might be better to be explicit in the config (with an enum or such), since at the end of the day, you will typically only be using one or the other? Then the user will also know which backend is used, and will not feel inclined to needlessly bundle both.

But I'm not of a strong opinion, this is likely only necessary during a transitional period, right? Can't imagine Dart going on forever without decent DTLS support :P

Reply:
> I think OpenSSL also has to be bundled for some platforms (iOS/Android), common targets for Flutter which also happen to be on metered connections where a smaller app size is an advantage. Not sure about Windows, but I think OpenSSL is not shipped with the OS there either?

Oh, yeah, that is a good point! And yes, OpenSSL has to be explicitly installed on Windows, if I am not mistaken.

> Might be better to be explicit in the config (with an enum or such), since at the end of the day, you will typically only be using one or the other? Then the user will also know which backend is used, and will not feel inclined to needlessly bundle both.

Sounds good, I updated the PR accordingly :)

> But I'm not of a strong opinion, this is likely only necessary during a transitional period, right? Can't imagine Dart going on forever without decent DTLS support :P

I really hope so. It is pretty interesting that DTLS seems to be such a challenge for many ecosystems (in Node.js, adding DTLS support to Node core has been discussed for years with no significant progress). I think adding it to the Dart SDK should be doable, though :) Especially with the work already done with the two backends present in this PR.